staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -649,48 +660,61 @@ func (alloc *allocator) allocateOne(r deviceIndices) (bool, error) {
 	claim := alloc.claimsToAllocate[r.claimIndex]
 	if r.requestIndex >= len(claim.Spec.Devices.Requests) {
 		// Done with the claim, continue with the next one.
-		return alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex + 1})
+		return alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex + 1}, false)
 	}
 
-	requestKey := requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: r.subRequest, subRequestIndex: r.subRequestIndex}
+	// r.subRequestIndex is zero unless the for loop below is the caller of allocateOne.
+	requestKey := requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequestIndex: r.subRequestIndex}
 	requestData := alloc.requestData[requestKey]
-	// If the request has subRequests, attempt them in order until one can be satisfied
-	// or all have been attempted.
-	if requestData.hasSubRequests {
-		for i := 0; ; i++ {
-			_, ok := alloc.requestData[requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: true, subRequestIndex: i}]
-			if !ok {
+
+	// Subrequests are special: we only need to allocate one of them, then
+	// we can move on to the next request. We enter this for loop when
+	// hitting the first subrequest, but not if we are already working on a
+	// specific subrequest.
+	if !allocateSubRequest && requestData.parentRequest != nil {
+		for subRequestIndex := 0; ; subRequestIndex++ {
+			nextSubRequestKey := requestKey
+			nextSubRequestKey.subRequestIndex = subRequestIndex
+			if _, ok := alloc.requestData[nextSubRequestKey]; !ok {
+				// Past the end of the subrequests without finding a solution -> give up.
 				return false, nil
 			}
-			success, err := alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: true, subRequestIndex: i})
+
+			r.subRequestIndex = subRequestIndex
+			success, err := alloc.allocateOne(r, true /* prevent infinite recusion */)
 			if err != nil {
 				return false, err
 			}
 			if success {
-				// Store the index of the selected subRequest
-				requestData.selectedSubRequestIndex = i
+				// Store the index of the selected subrequest.
+				requestData.selectedSubRequestIndex = subRequestIndex
 				alloc.requestData[requestKey] = requestData
 				return true, nil
 			}
 		}
+		// No success.
+		return false, nil
```

Comment:
Wasn't this return missing in your code? The test cases pass with and without it, which indicates that we miss some test case where it is relevant.

I added it while writing the for loop from scratch and only now when looking at the diff noticed that you didn't have it.